### PR TITLE
Remove bootstrapping code

### LINKED ISSUE #28: PutValue context not utilized or propagated into bootstrapping
`PubsubValueStore.PutValue` receives a `context.Context` but does not respect it. In particular, we have: https://github.com/libp2p/go-libp2p-pubsub-router/blob/2c8ac20e642241b8233dbfa5bcb307c8997499b6/pubsub.go#L92 

which uses the `PubsubValueStore`'s context instead of the `PutValue` context when calling the very expensive `bootstrapPubsub` function.

Additionally, `bootstrapPubsub` could be much less expensive if:
https://github.com/libp2p/go-libp2p-pubsub-router/blob/2c8ac20e642241b8233dbfa5bcb307c8997499b6/pubsub.go#L345

was moved into the go func right below it.

After speaking with @Stebalien my current plan is to start by moving the `Provide` into the go func, and then look into moving bootstrapping into PubSub via the Discovery (https://github.com/libp2p/go-libp2p-discovery) interface, as indicated by: https://github.com/libp2p/go-libp2p-pubsub-router/blob/2c8ac20e642241b8233dbfa5bcb307c8997499b6/pubsub.go#L338 

@vyzo @raulk Any thoughts or issues with this?

### PR DESCRIPTION
**Pull request by aschmahmann**
Now that PubSub has an optional discovery service built into it seems reasonable for the PubsubValueStore to assume that the PubSub instance passed into it utilizes this discovery mechanism (if that is in fact what the caller intends). This closes #28.

While we're doing breaking the constructor and semantics anyway (removing the mandatory ContentRouter and needing the PubSub instance passed in to have Discovery enabled) I also added an option pattern to the constructor.

### CODE REVIEW

File: pubsub.go [hunk 1]
```diff
@@ -162,33 +182,47 @@ func (p *PubsubValueStore) Subscribe(key string) error {
 		return cmp > 0 || cmp == 0 && src == myID
 	})
 
-	sub, err := p.ps.Subscribe(topic)
-	if err != nil {
+	if err := p.fillTopicInfo(ti, topic); err != nil {
```
Comment (Stebalien) [inline pubsub.go]: We're not holding the lock.

File: pubsub.go [hunk 2]
```diff
@@ -162,33 +182,47 @@ func (p *PubsubValueStore) Subscribe(key string) error {
...
+	if err != nil {
+		if closeErr := t.Close(); closeErr != nil {
+			return multierror.Append(err, closeErr)
+		}
+		return err
 	}
 
+	evts, err := t.EventHandler()
+	if err != nil {
+		sub.Cancel()
```
Comment (Stebalien) [inline pubsub.go]: Will this not be automatically closed when we close the handle?

File: pubsub.go [hunk 3]
```diff
@@ -373,15 +413,23 @@ func (p *PubsubValueStore) Cancel(name string) (bool, error) {
...
+func (p *PubsubValueStore) handleSubscription(ctx context.Context, ti *topicInfo, key string, cancel func()) {
+	defer func() {
+		if err := ti.topic.Close(); err != nil {
+			if err != context.Canceled && err != context.DeadlineExceeded {
+				log.Error(err)
+			}
+		}
+	}()
+	defer ti.evts.Cancel()
+	defer ti.sub.Cancel()
```
Comment (Stebalien) [inline pubsub.go]: Do we need to explicitly close these?

File: pubsub.go (hunk 1 above)
Comment (aschmahmann) [reply pubsub.go]: Might as well to be safe. I think I was relying on p.ps.Join only succeeding the first time it's called (i.e. only 1 Topic can exist per topic). This is probably not something we should overly rely on though.

File: pubsub.go (hunk 2 above)
Comment (aschmahmann) [reply pubsub.go]: It's currently setup so that you can't close the topic handle if there are existing subscriptions or event handlers (i.e. it returns "cannot close topic: outstanding event handlers or subscriptions")

File: pubsub.go (hunk 3 above)
Comment (aschmahmann) [reply pubsub.go]: It's probably not a big deal the ways this function exits are:
1. The PubSubValueStore's context is cancelled
2. The internal subscription was manually cancelled

Given that the second shouldn't occur outside of this code (and if it does it's probably unsafe) and the first means the whole service is dead anyway it's probably fine to leave it. Was keeping it mostly for consistency with the previous system, although probably having an internal `Close` function as in https://github.com/libp2p/go-libp2p-pubsub-router/issues/35 is probably better.

I can look into it.

File: pubsub.go (hunk 2 above)
Comment (Stebalien) [reply pubsub.go]: Hm. I'd expect to be able to just close the topic and walk away. Could you file an issue? We can fix this later.

File: pubsub.go (hunk 1 above)
Comment (Stebalien) [reply pubsub.go]: Got it. Won't rebroadcast concurrently read this?

Comment (aschmahmann) [reply pubsub.go]: probably, better off protecting this anyway.

File: pubsub.go (hunk 2 above)
Comment (aschmahmann) [reply pubsub.go]: We use a similar pattern here where you can't call `Close(topic)` on the PubSubValueStore if SearchValue is ongoing. Should I file an issue here as well?

Comment (Stebalien) [reply pubsub.go]: These are two very different APIs:

* In `PubSub`, `Topics` handles are _owned_. If I call `Close` on a topic, I'm not affecting any other services. That's why I was so adamant that we either:
  * Only allowing one service to use a topic at a time (current implementation).
  * Make topic handles completely independent.
* In the `PubSubValueStore`, multiple unrelated services might be independently calling `SearchValue`. I don't want some other service to come along and cancel these searches.

File: pubsub.go (hunk 3 above)
Comment (aschmahmann) [reply pubsub.go]: The cancels are idempotent so it shouldn't be a big deal, also I took a stab at fixing #35 as part of cleaning this up. lmk what you think.

File: pubsub.go (hunk 1 above)
Comment (aschmahmann) [reply pubsub.go]: better?

File: pubsub.go (hunk 2 above)
Comment (aschmahmann) [reply pubsub.go]: Issue filed libp2p/go-libp2p-pubsub#237

File: pubsub.go [hunk 4]
```diff
@@ -42,71 +42,95 @@ type PubsubValueStore struct {
...
+	p.mx.Unlock()
+
+	ti.dbWriteMx.Lock()
+	defer ti.dbWriteMx.Unlock()
+	ok, err := p.putLocal(ti, key, value)
+	if err != nil {
+		return err
+	}
+	if !ok {
+		return nil
```
Comment (aschmahmann) [inline pubsub.go]: Is `return nil` better than something like `return errors.New("put record that is worse than the current state")`?

File: fetch_test.go [hunk 5]
```diff
@@ -46,8 +46,8 @@ func TestFetchProtocolTrip(t *testing.T) {
 	d2 := &datastore{map[string][]byte{"key": []byte("value2")}}
 	h2 := newFetchProtocol(ctx, hosts[1], d2.Lookup)
 
-	fetchCheck(t, ctx, h1, h2, "key", []byte("value2"))
-	fetchCheck(t, ctx, h2, h1, "key", []byte("value1"))
+	fetchCheck(ctx, t, h1, h2, "key", []byte("value2"))
```
Comment (aschmahmann) [inline fetch_test.go]: Could move this refactor to a separate PR if we wanted, but it's pretty simple.

File: pubsub.go [hunk 6]
```diff
@@ -42,71 +42,95 @@ type PubsubValueStore struct {
...
+	ti, ok := p.topics[key]
+	if !ok {
+		p.mx.Unlock()
+		return errors.New("could not find topic handle")
+	}
+	p.mx.Unlock()
+
+	ti.dbWriteMx.Lock()
+	defer ti.dbWriteMx.Unlock()
+	ok, err := p.putLocal(ti, key, value)
```
Comment (aschmahmann) [inline pubsub.go]: Previously we were receiving updates by waiting for `pubsub.Publish(topic,value)` to return us a value via our own `Subscription`. However, pubsub can drop messages so this isn't great, we now actually do a local put before the pubsub Put.

This does mean that we'll receive most of our messages twice (and therefore deserialize, validate and compare them to the most recent state twice).

File: pubsub.go [hunk 7]
```diff
@@ -359,29 +411,49 @@ func (p *PubsubValueStore) Cancel(name string) (bool, error) {
...
-	defer cancel()
+func (p *PubsubValueStore) handleSubscription(ctx context.Context, ti *topicInfo, key string) {
+	defer func() {
+		ti.finished <- struct{}{}
+
+		p.mx.Lock()
+		defer p.mx.Unlock()
+
+		if err := p.closeTopic(key, ti); err != nil {
+			if err != context.Canceled && err != context.DeadlineExceeded {
```
Comment (aschmahmann) [inline pubsub.go]: doing a type check here for the error doesn't feel great (what if pubsub decided to start wrapping errors, we'd have to pay attention). However, since there's nothing to do but log the messages this isn't too bad.

Unfortunately, since the pattern we use to deconstruct/close libp2p components is to have their parent context's expire and we generally have one cancellable parent context per application it's possible PubSub will get cancelled before the PubsubValueStore does.

File: pubsub.go [hunk 8]
```diff
@@ -359,29 +411,49 @@ func (p *PubsubValueStore) Cancel(name string) (bool, error) {
...
+func (p *PubsubValueStore) closeTopic(key string, ti *topicInfo) error {
+	ti.cancel()
+	ti.sub.Cancel()
+	ti.evts.Cancel()
+	if err := ti.topic.Close(); err != nil {
+		return err
 	}
 
-	return ok, nil
+	delete(p.topics, key)
```
Comment (Stebalien) [inline pubsub.go]: We should do this regardless.

File: fetch_test.go (hunk 5 above)
Comment (Stebalien) [reply fetch_test.go]: Np. Small cleanups are almost always fine to just include.

File: pubsub.go [hunk 9]
```diff
@@ -42,71 +42,95 @@ type PubsubValueStore struct {
...
 	// Record-store keys are arbitrary binary. However, pubsub requires UTF-8 string topic IDs.
 	// Encodes to "/record/base64url(key)"
 	return "/record/" + base64.RawURLEncoding.EncodeToString([]byte(key))
 }
 
-// NewPubsubPublisher constructs a new Publisher that publishes IPNS records through pubsub.
-// The constructor interface is complicated by the need to bootstrap the pubsub topic.
-// This could be greatly simplified if the pubsub implementation handled bootstrap itself
-func NewPubsubValueStore(ctx context.Context, host host.Host, cr routing.ContentRouting, ps *pubsub.PubSub, validator record.Validator) *PubsubValueStore {
+type PubsubValueStoreOpt func(*PubsubValueStore) error
```
Comment (Stebalien) [inline pubsub.go]: Nit: might as well just call this `Option`. (and we should document it).

File: pubsub.go [hunk 10]
```diff
@@ -42,71 +42,95 @@ type PubsubValueStore struct {
...
 	if err := p.Subscribe(key); err != nil {
 		return err
 	}
 
 	log.Debugf("PubsubPublish: publish value for key", key)
 
+	p.mx.Lock()
+	ti, ok := p.topics[key]
+	if !ok {
+		p.mx.Unlock()
```
Comment (Stebalien) [inline pubsub.go]: Just put this on line 116 and unlock once?

File: pubsub.go (hunk 4 above)
Comment (Stebalien) [reply pubsub.go]: This is wrong. We need to publish regardless because both the DHT and this package will fight over updating the datastore.

Comment (Stebalien) [reply pubsub.go]: We should check to see if the new value is _at least_ as new as the local value.

File: pubsub.go (hunk 6 above)
Comment (Stebalien) [reply pubsub.go]: We shouldn't be dropping locally published messages. But yeah, it makes sense to put them locally anyways just in case.

File: pubsub.go [hunk 11]
```diff
@@ -359,29 +411,49 @@ func (p *PubsubValueStore) Cancel(name string) (bool, error) {
...
+	delete(p.topics, key)
+	return nil
 }
 
-func (p *PubsubValueStore) handleSubscription(ctx context.Context, sub *pubsub.Subscription, key string, cancel func()) {
-	defer sub.Cancel()
-	defer cancel()
+func (p *PubsubValueStore) handleSubscription(ctx context.Context, ti *topicInfo, key string) {
+	defer func() {
+		ti.finished <- struct{}{}
```
Comment (Stebalien) [inline pubsub.go]: The idiom is to close signal channels. That way, waiting twice for some reason (e.g., if closing the topic fails for some reason) doesn't break anything).

File: pubsub.go (hunk 7 above)
Comment (Stebalien) [reply pubsub.go]: We can just check if the context has been canceled. If `ctx.Err() != nil`, we're shutting down and might as well ignore any errors.

Comment (Stebalien) [reply pubsub.go]: Wait, why is `closeTopic` returning errors _anyways_? Even when we're shutting down, `Close` should just close and go away.

File: pubsub.go (hunk 1 above)
Comment (Stebalien) [reply pubsub.go]: Note: 

> probably, better off protecting this either way.

This is _never_ a good idea. If you don't know whether or not a lock is necessary, the code is probably buggy.

File: pubsub.go (hunk 4 above)
Comment (aschmahmann) [reply pubsub.go]: > We should check to see if the new value is at least as new as the local value.

putLocal checks to see if the new value is **better** (for IPNS ~= newer) than the local value.

> We need to publish regardless because both the DHT and this package will fight over updating the datastore.

I'm confused by what you mean here, AFAIK here are the records used by IPNS:

- Namesys: `/ipns/base32(binaryPeerID)`
- DHT: `base32(/ipns/binaryPeerID)`
- PubSub Router (here): `/base32(/ipns/binaryPeerID)`
   - Up until this PR was only storing the data in a local map datastore, so there was no "fighting"
   - This PR adds an option to use another datastore (e.g. the main one) so that we can eventually move to having a shared datastore (in order to enable third party republishing)
        - I can just remove the datastore option for now and then no one will possibly mess this up. We can have a separate PR that adds the datastore option and changes the record to look like the Namesys record.

File: pubsub.go [hunk 12]
```diff
@@ -206,19 +245,21 @@ func (p *PubsubValueStore) rebroadcast(ctx context.Context) {
...
-				for _, k := range keys {
+			if len(topics) > 0 {
+				for i, k := range keys {
 					val, err := p.getLocal(k)
 					if err == nil {
-						topic := KeyToTopic(k)
+						topic := topics[i].topic
 						select {
-						case <-p.psPublishChannel(topic, val):
+						case <-p.psPublishChannel(ctx, topic, val):
```
Comment (aschmahmann) [inline pubsub.go]: In theory we could emit an old record here and we should use a mutex or something to protect it. However, emitting old records isn't so bad and the network won't degrade much assuming infrequent rebroadcasts (which should be how this is used).

I'm also hoping we can largely just remove rebroadcasting entirely when PubSub starts emitting events for dropped messages.

I'm trying to minimize the number of changes that need review to get this approved, but if you'd like me to fix this lmk and I'll get right on it.

File: pubsub.go [hunk 13]
```diff
@@ -206,19 +245,21 @@ func (p *PubsubValueStore) rebroadcast(ctx context.Context) {
 		select {
 		case <-ticker.C:
 			var keys []string
+			var topics []*topicInfo
 			p.mx.Lock()
-			keys = make([]string, 0, len(p.subs))
-			for p, _ := range p.subs {
-				keys = append(keys, p)
+			topics = make([]*topicInfo, 0, len(p.topics))
```
[the remark this answers is not in the document]
Comment (Stebalien) [reply pubsub.go]: > so we don't need to hold p.mx during this loop

I meant forward declaring here: https://github.com/libp2p/go-libp2p-pubsub-router/pull/37/commits/f0530f5fee2aabda9eba2bc29cecfa6fdd0388c4#diff-16720ecd1b41b213c82a47c8271e3096L247-L248

File: pubsub.go (hunk 4 above)
Comment (Stebalien) [reply pubsub.go]: Got it. I assumed the datastore was shared.

Let's make the check "at least as good as" anyways. That's what we do in the DHT and, as you say, it's what we'll need to do if we share the datastore.

File: pubsub.go (hunk 12 above)
Comment (Stebalien) [reply pubsub.go]: Ah, you mean, given a concurrent publish, we could:

1. Start a rebroadcast tick.
2. Get an old value with `getLocal`.
3. Publish a new value (independently).
4. Re-publish the old value?

Yeah, that sounds fine. We could also just filter out strictly _worse_ values in the validator.

(FYI, providing event sequences like that makes understanding issues like this much easier. "This could emit old an old record" doesn't tell me how.)

File: pubsub.go (hunk 8 above)
Comment (Stebalien) [reply pubsub.go]: I'm generally not a fan of ignoring errors from anywhere, but this is probably fine.

File: pubsub.go (hunk 4 above)
Comment (aschmahmann) [reply pubsub.go]: > Let's make the check "at least as good as" anyways.

I'm confused, I already do that with the boolean returned by `putLocal` anyway. No?

File: pubsub.go (hunk 7 above)
Comment (Stebalien) [reply pubsub.go]: Ignoring it is probably fine.

Comment (aschmahmann) [reply pubsub.go]: done through this file

File: pubsub.go [hunk 14]
```diff
@@ -162,34 +183,49 @@ func (p *PubsubValueStore) Subscribe(key string) error {
...
-	if existingSub != nil {
-		p.mx.Unlock()
-		sub.Cancel()
-		return nil
-	}
-
+	p.topics[key] = ti
 	ctx, cancel := context.WithCancel(p.ctx)
-	p.cancels[key] = cancel
+	ti.cancel = cancel
```
Comment (aschmahmann) [inline pubsub.go]: kind of annoying that this isn't handled inside `createTopicHandler`, but it's easier to deal with the contexts this way

File: pubsub.go (hunk 4 above)
Comment (Stebalien) [reply pubsub.go]: We're currently checking if the new value is strictly better than the existing value. We should be checking that the new value is at least as good as the new value. That way:

1. The user can explicitly re-publish an existing value.
2. If/when we start sharing a datastore with other routers, we won't see the newer value from the DHT and then decide to _not_ broadcast our value over pubsub.

Comment (aschmahmann) [reply pubsub.go]: Ah, thanks for clarifying. Not an unreasonable request. lmk if it's good enough now.

**Comment by aschmahmann** [comment]
Before I merge this it'd be nice to be able to run CI on the [go-ipfs PR](https://github.com/ipfs/go-ipfs/pull/6758) that incorporates the change, for extra testing. However, that's currently blocked bc this repo was updated to the latest version of the datastore interface.

@Stebalien should I just merge this and we can worry about the integration later?

File: pubsub_test.go [hunk 15]
```diff
@@ -309,6 +315,69 @@ func TestWatch(t *testing.T) {
...
+	const baseKey = "/namespace/key"
+
+	for i := 0; i < numRuns; i++ {
+		key := fmt.Sprintf("%s/%d", baseKey, i)
+		doneCh := make(chan error, numRoutines)
+		for j := 0; j < numRoutines; j++ {
+			rtNum := j
+			go func() {
+				err := vss[0].PutValue(ctx, key, []byte(fmt.Sprintf(fmtString, key, rtNum)))
+				doneCh <- err
```
Comment (dirkmc) [inline pubsub_test.go]: It may be possible to simplify this somewhat using an [errgroup](https://godoc.org/golang.org/x/sync/errgroup)

Comment (aschmahmann) [reply pubsub_test.go]: done
